packages/core/src/State.ts

```diff
@@ -298,7 +300,9 @@ export class State<
    * @param parentStateValue
    */
   public matches<TSV extends TTypestate['value']>(
-    parentStateValue: TSV
+    parentStateValue: TMeta extends IsTypegenActive
+      ? TMeta['matchesStates']
+      : TSV
```

Comment:
I had to do a bit of trampling over `TTypestate` here - basically overriding the inference in favour of the meta info we get from `TMeta['matchesStates']`